Switch to Pkg mode prompt immediately and load Pkg in the background

IanButterworth opened the pull request:
Because Pkg is now a pkgimage and can load slowly on slower machines, which is a bit frustrating in the repl.

This makes the repl immediately switch to a dummy prompt that looks like Pkg mode to allow the user to keep typing while Pkg loads. During which the keymap is disabled.

It works best if julia has >1 thread, otherwise typing stalls during Pkg load.
Side note, it might make sense if by default julia in interactive mode started with at least `-t1,1`.

If Pkg takes longer to load than the user to type the command and press return, then the UX isn't great as it won't do anything.

Opening as a draft as the threading limitation is a bit awkward, and there may be a better approach to achieve the same UX.



https://github.com/JuliaLang/julia/assets/1694067/1bf17323-441a-4db2-8a3b-4d571eac622f

IanButterworth committed:
switch to pkg mode prompt immediately

gbaraldi commented on stdlib/REPL/src/REPL.jl:
```diff
@@ -1230,22 +1230,35 @@ function setup_interface(
+                        pkgid = Base.PkgId(Base.UUID("44cfe95a-1eb2-52ea-b672-e2afdf69b78f"), "Pkg")
+                        Base.require_stdlib(pkgid, "REPLExt")
+                    end
+                    t_replswitch = @async begin
```
Why not `@spawn`?

IanButterworth replied:
I had an earlier design that required two tasks, but I think you're right, and this can be simplified into a single spawn now

gbaraldi replied:
Even still, `@async` should probably never be used because it does bad things

IanButterworth replied:
What if you want a task that strictly stays on the same thread as the task it was created from? Is that just a bad pattern?

gbaraldi replied:
Yep, because threads aren't really concept that exist in julia, so it's basically an implementation detail.

IanButterworth replied:
I think you mean `@async`, but thanks for the explanation. 

Btw, maybe we should just go ahead and deprecate the Threads module to Tasks. Perhaps without a warning though, so just the alias and a docs note.

gbaraldi replied:
Yeah, that name isn't very helpful.

IanButterworth commented:
I think even without >1 threads this is an improvement in UX responsiveness, so it should be good to go.

IanButterworth committed:
revert pkg_mode caching & tidy

Reverts "Pkg REPL: cache `pkg_mode` lookup (#54359)"
as this function is replaced when Pkg loads by a simpler repl switch so no caching is needed

IanButterworth committed:
don't take from stdin if loading is finished

IanButterworth committed:
check for hint after repl switch

IanButterworth committed:
reduce chance of race

IanButterworth committed:
add some precompiles

IanButterworth committed:
only check for hint if prompt is not empty

christiangnrd commented:
This is great and makes a significant difference in how it feels to use the repl in master!

However, I noticed an issue that I demonstrate in the video. If you press enter before Pkg finishes loading (very possible when you want to start by checking the status of the current environment), nothing happens, and when you press enter again, it errors.

https://github.com/JuliaLang/julia/assets/28689358/5db69bb7-0e9f-4d7b-9c50-9cb8f0c70a8c

IanButterworth commented:
Thanks for the quick report, I can reproduce this. It's because we need to ignore enters chars in the stdin capture during the load so that they don't get entered into the pkg prompt buffer, which doesn't usually happen.

Can you confirm that https://github.com/JuliaLang/julia/pull/54674 fixes this for you. Thanks

christiangnrd commented:
I'll respond in the new PR.

I also managed to cause some funky display behaviour when testing out 13635e1, but I can't remember how I did it. I'll report back if I figure it out.
